keycloak/client.go

```diff
@@ -0,0 +1,202 @@
+	AddUserToGroup(ctx context.Context, token, realm, userID, groupID string) error
+	DeleteUserFromGroup(ctx context.Context, token, realm, userID, groupID string) error
+}
+
+// Client interacts with the Keycloak API
+type Client struct {
+	Client GoCloak
+
+	Realm    string
+	Username string
+	Password string
+}
+
+// NewClient creates a new Client
+func NewClient(host, realm, username, password string) Client {
+	return Client{
+		Client:   gocloak.NewClient(host),
+		Realm:    realm,
+		Username: username,
+		Password: password,
+	}
+}
+
+// PutGroup creates the provided Keycloak group if it does not exist and adjusts the group members accordingly.
+// The method is idempotent.
+func (c Client) PutGroup(ctx context.Context, group Group) (Group, error) {
+	res := Group{
+		Name: group.Name,
+	}
+
+	token, err := c.Client.LoginAdmin(ctx, c.Username, c.Password, c.Realm)
+	if err != nil {
+		return res, fmt.Errorf("failed binding to keycloak: %w", err)
+	}
+
+	found, foundMemb, err := c.getGroupAndMembersByName(ctx, token, group.Name)
+	if err != nil {
+		return res, fmt.Errorf("failed finding group: %w", err)
+	}
+	if found == nil {
+		id, err := c.Client.CreateGroup(ctx, token.AccessToken, c.Realm, gocloak.Group{
+			Name: gocloak.StringP(group.Name),
+		})
+		if err != nil {
+			return res, err
+		}
+		found = &gocloak.Group{
+			ID:   &id,
+			Name: gocloak.StringP(group.Name),
+		}
+	}
+
+	for _, fm := range foundMemb {
+		if !contains(group.Members, *fm.Username) {
+			// user is not in group remove it
+			err := c.Client.DeleteUserFromGroup(ctx, token.AccessToken, c.Realm, *fm.ID, *found.ID)
+			if err != nil {
+				return res, err
+			}
+		} else {
+			res.Members = append(res.Members, *fm.Username)
+		}
+	}
+	newMemb := diff(group.Members, res.Members)
+
+	addedMemb, err := c.addUsersToGroup(ctx, token, *found.ID, newMemb)
+	res.Members = append(res.Members, addedMemb...)
+	return res, err
+}
+
+// DeleteGroup deletes the Keycloak group by name.
+// The method is idempotent and will not do anything if the group does not exits.
+func (c Client) DeleteGroup(ctx context.Context, groupName string) error {
+	token, err := c.Client.LoginAdmin(ctx, c.Username, c.Password, c.Realm)
+	if err != nil {
+		return fmt.Errorf("failed binding to keycloak: %w", err)
+	}
+	found, err := c.getGroupByName(ctx, token, groupName)
+	if err != nil {
+		return fmt.Errorf("failed finding group: %w", err)
+	}
+	if found == nil {
+		return nil
+	}
+	return c.Client.DeleteGroup(ctx, token.AccessToken, c.Realm, *found.ID)
+}
+
+func (c Client) getGroupByName(ctx context.Context, token *gocloak.JWT, name string) (*gocloak.Group, error) {
+	// This may return more than one 1 result
+	groups, err := c.Client.GetGroups(ctx, token.AccessToken, c.Realm, gocloak.GetGroupsParams{
+		Search: &name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var group *gocloak.Group
+	for i := range groups {
+		if *groups[i].Name == name {
+			group = groups[i]
```

Comment:
Minor nitpick: should we break out of the loop here?

```suggestion
			group = groups[i]
			break;
```

Or directly return, like we do in `getUserByName()`

```suggestion
			return groups[i], nil
```

Reply:
Good point. That should work the same way as in `getUserByName` for consistency.